File: drills/python/harnesses/nc150_serialize_deserialize.py
try:
    (serialize, deserialize, TreeNode)
except NameError:
    raise Exception("__signature_mismatch__")

__results__ = []
def __case__(label, expected, actual):
    __results__.append([label, repr(expected), repr(actual)])

# LeetCode's level-order form: a list of values with None for a missing child,
# trailing Nones trimmed.
def __build__(values):
    if not values or values[0] is None:
        return None
    root = TreeNode(values[0])
    queue, head, i = [root], 0, 1
    while head < len(queue) and i < len(values):
        node = queue[head]
        head += 1
        if i < len(values):
            value = values[i]
            i += 1
            if value is not None:
                node.left = TreeNode(value)
                queue.append(node.left)
        if i < len(values):
            value = values[i]
            i += 1
            if value is not None:
                node.right = TreeNode(value)
                queue.append(node.right)
    return root


def __levels__(node):
    out, queue, head = [], [node], 0
    while head < len(queue) and len(out) < 500:
        current = queue[head]
        head += 1
        if current is None:
            out.append(None)
        else:
            out.append(current.val)
            queue.append(current.left)
            queue.append(current.right)
    while out and out[-1] is None:
        out.pop()
    return out

# The format is free, so what is checked is the round trip.
def __round_trip__(values):
    return __levels__(deserialize(serialize(__build__(values))))

__case__("deserialize(serialize([1,2,3,None,None,4,5]))", [1, 2, 3, None, None, 4, 5], __round_trip__([1, 2, 3, None, None, 4, 5]))
__case__("deserialize(serialize([]))", [], __round_trip__([]))
__case__("deserialize(serialize([0]))", [0], __round_trip__([0]))
__case__("deserialize(serialize(a lopsided tree))", [1, 2, None, 3, None, None, 4], __round_trip__([1, 2, None, 3, None, None, 4]))
__case__("deserialize(serialize([-1,-2,-3])) -- negatives survive", [-1, -2, -3], __round_trip__([-1, -2, -3]))

File: drills/ts/harnesses/nc150_serialize_deserialize.ts
import * as solution from "./solution";

const show = (v: unknown) => JSON.stringify(v) ?? "undefined";

// LeetCode's level-order form: an array of values with null for a missing
// child, trailing nulls trimmed.
const build = (values: (number | null)[]): any => {
  if (values.length === 0 || values[0] === null) return null;
  const root = new solution.TreeNode(values[0]!);
  const queue: any[] = [root];
  let head = 0;
  let i = 1;
  while (head < queue.length && i < values.length) {
    const node = queue[head++];
    if (i < values.length) {
      const value = values[i++];
      if (value !== null) {
        node.left = new solution.TreeNode(value);
        queue.push(node.left);
      }
    }
    if (i < values.length) {
      const value = values[i++];
      if (value !== null) {
        node.right = new solution.TreeNode(value);
        queue.push(node.right);
      }
    }
  }
  return root;
};

const levels = (node: any): (number | null)[] => {
  const out: (number | null)[] = [];
  const queue: any[] = [node];
  let head = 0;
  while (head < queue.length && out.length < 500) {
    const current = queue[head++];
    if (current === null || current === undefined) {
      out.push(null);
    } else {
      out.push(current.val);
      queue.push(current.left ?? null);
      queue.push(current.right ?? null);
    }
  }
  while (out.length > 0 && out[out.length - 1] === null) out.pop();
  return out;
};

// The format is free, so what is checked is the round trip.
const roundTrip = (values: (number | null)[]) =>
  levels(solution.deserialize(solution.serialize(build(values))));

export function run(): [string, string, string][] {
  if (typeof solution.serialize !== "function" || typeof solution.deserialize !== "function" || typeof solution.TreeNode !== "function") throw new Error("__signature_mismatch__");
  return [
    ["deserialize(serialize([1,2,3,null,null,4,5]))", show([1, 2, 3, null, null, 4, 5]), show(roundTrip([1, 2, 3, null, null, 4, 5]))],
    ["deserialize(serialize([]))", show([]), show(roundTrip([]))],
    ["deserialize(serialize([0]))", show([0]), show(roundTrip([0]))],
    ["deserialize(serialize(a lopsided tree))", show([1, 2, null, 3, null, null, 4]), show(roundTrip([1, 2, null, 3, null, null, 4]))],
    ["deserialize(serialize([-1,-2,-3])) -- negatives survive", show([-1, -2, -3]), show(roundTrip([-1, -2, -3]))],
  ];
}
